Add extensions feature category and flag for extension sidebar APIs

AriYoung00 opened the pull request:
#### 17811113e0171723d2f52ce0f53f3793e4225818
<pre>
Add extensions feature category and flag for extension sidebar APIs

<a href="https://bugs.webkit.org/show_bug.cgi?id=276551">https://bugs.webkit.org/show_bug.cgi?id=276551</a>
<a href="https://rdar.apple.com/131476290">rdar://131476290</a>

Reviewed by Brian Weinstein and Timothy Hatcher.

This PR adds a feature flag to control the enablement of the sidePanel
and sidebarAction WebExtension APIs. Additionally, this PR also adds a
compile-time switch to control the enablement of these APIs when
compiling. Finally, this PR also adds a category for &quot;Extensions&quot; to the
feature flag preferences.

* Source/WTF/Scripts/GeneratePreferences.rb: Add &apos;Extensions&apos; category
  to list of valid feature flag preference categories.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml: Add feature
  flag and defaults to web preferences (conditional on compile-time
  flag).
* Source/WTF/wtf/PlatformEnableCocoa.h: Add compile time flag
  ENABLE_WK_WEB_EXTENSIONS_SIDEBAR
* Source/WebKit/UIProcess/API/APIFeatureStatus.h: Add `Extensions` to
  `FeatureCategory` enum.
* Source/WebKit/UIProcess/API/Cocoa/_WKFeature.mm:
(-[_WKFeature category]): Add switch case to `category` property mapping
  `API::FeatureCategory::Extensions` to `WebFeatureCategoryExtensions`.
* Source/WebKit/UIProcess/API/glib/WebKitFeature.cpp:
(toFeatureCategory): Add switch case mapping
  `API::FeatureCategory::Extensions` to `&quot;Extensions&quot;`.
* Source/WebKitLegacy/mac/WebView/WebFeature.h: Add
  `WebFeatureCategoryExtensions` to `WebFeatureCategory` enum.

Canonical link: <a href="https://commits.webkit.org/281065@main">https://commits.webkit.org/281065@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e546a6e1c26fb427f8061ceea4f39ceb0dc77d6c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58461 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37788 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10945 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62087 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8905 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45424 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9102 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/47342 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6349 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60492 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35388 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50545 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28187 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32140 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7847 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7909 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/51552 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54093 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8118 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63790 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/57703 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2374 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8140 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54664 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2382 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50571 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54730 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12938 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2009 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/79463 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/33617 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/13224 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34703 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35787 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/34448 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKitLegacy/mac/WebView/WebFeature.h:
```diff
@@ -64,7 +64,8 @@ typedef NS_ENUM(NSUInteger, WebFeatureCategory) {
     WebFeatureCategorySecurity = 8,
-    WebFeatureCategoryHTML = 9
+    WebFeatureCategoryHTML = 9,
+    WebFeatureCategoryExtensions = 10
```
`WebFeatureCategoryExtensions = 10,` to prevent someone else having to change this line when another category is added

AriYoung00 replied:
Alrighty, this should be updated with the latest changes -- thanks!

xeenon commented on Source/WTF/wtf/PlatformEnableCocoa.h:
```diff
@@ -1002,6 +1002,7 @@
 
 #if !defined(ENABLE_WK_WEB_EXTENSIONS) && (PLATFORM(MAC) || PLATFORM(MACCATALYST) || PLATFORM(IOS) || PLATFORM(IOS_SIMULATOR) || PLATFORM(VISION))
 #define ENABLE_WK_WEB_EXTENSIONS 1
+#define ENABLE_WK_WEB_EXTENSIONS_SIDEBAR 0
```
This likely is best to be broken out on its own and be set to:

```
#if !defined(ENABLE_WK_WEB_EXTENSIONS_SIDEBAR)
#define ENABLE_WK_WEB_EXTENSIONS_SIDEBAR 0 && ENABLE_WK_WEB_EXTENSIONS
#endif
```

AriYoung00 replied:
Alrighty, this change should be in the PR now -- thanks!

AriYoung00 committed:
Add extensions feature category and flag for extension sidebar APIs

https://bugs.webkit.org/show_bug.cgi?id=276551
rdar://131476290

Reviewed by Brian Weinstein and Timothy Hatcher.

This PR adds a feature flag to control the enablement of the sidePanel
and sidebarAction WebExtension APIs. Additionally, this PR also adds a
compile-time switch to control the enablement of these APIs when
compiling. Finally, this PR also adds a category for "Extensions" to the
feature flag preferences.

* Source/WTF/Scripts/GeneratePreferences.rb: Add 'Extensions' category
  to list of valid feature flag preference categories.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml: Add feature
  flag and defaults to web preferences (conditional on compile-time
  flag).
* Source/WTF/wtf/PlatformEnableCocoa.h: Add compile time flag
  ENABLE_WK_WEB_EXTENSIONS_SIDEBAR
* Source/WebKit/UIProcess/API/APIFeatureStatus.h: Add `Extensions` to
  `FeatureCategory` enum.
* Source/WebKit/UIProcess/API/Cocoa/_WKFeature.mm:
(-[_WKFeature category]): Add switch case to `category` property mapping
  `API::FeatureCategory::Extensions` to `WebFeatureCategoryExtensions`.
* Source/WebKit/UIProcess/API/glib/WebKitFeature.cpp:
(toFeatureCategory): Add switch case mapping
  `API::FeatureCategory::Extensions` to `"Extensions"`.
* Source/WebKitLegacy/mac/WebView/WebFeature.h: Add
  `WebFeatureCategoryExtensions` to `WebFeatureCategory` enum.

Canonical link: https://commits.webkit.org/281065@main
